### ISSUE TITLE
FNM1D should update at least once

### ISSUE THREAD

**Issue by mdpiper** [permamodel/permamodel#44]
The CSDMS Modeling Framework expects a model to update at least once, as a part of the initialize-run-finalize lifecycle.

Using the default configuration file produced from its BMI settings (see below), FNM1D throws an exception when updated. For example:

```
In [1]: from permamodel.components.bmi_frost_number import BmiFrostnumberMethod
In [2]: x = BmiFrostnumberMethod()
In [3]: x.initialize('frostnumber_model.cfg')
In [4]: x._model._configuration['start_year']
Out[4]: 2000
In [5]: x._model._configuration['end_year']
Out[5]: 2000
In [6]: x.get_value('frostnumber__air')
Out[6]: array(0.3961466061143555)
In [7]: x.update()
---------------------------------------------------------------------------
<snip>
ValueError: Year is past last year
```

It would be better if the value of `frostnumber__air` was set to some default value (e.g., -999.0) on initialization. The value of `frostnumber__air` for the one year run of the model would then be calculated during the update step. For example:

```
In [10]: x.initialize('frostnumber_model.cfg')
In [11]: x.get_value('frostnumber__air')
Out[11]: array(-999.0)
In [12]: x.update()
In [12]: x.get_value('frostnumber__air')
Out[12]: array(0.3961466061143555)
```

This is something that's not documented well by CSDMS, and something that I haven't explained well.

```
#===============================================================================
# PermaModel Config File for: frostnumber_method
#===============================================================================
# Input
permafrost_dir  | .          | string   | Set to root dir
comp_status     | Enabled             | string   | component status {Enabled; Disabled}
in_directory    | .            | string   | input directory
out_directory   | .           | string   | output directory
site_prefix     | Test             | string   | file prefix for the study site
case_prefix     | Case_test             | string   | file prefix for the model scenario
T_air_min_type  | Scalar          | string               | allowed: {Scalar; Time_Series}
T_air_min       | -9.0               | float   | Mean month air temperature coldest month [C]
T_air_max_type  | Scalar          | string               | allowed: {Scalar; Time_Series}
T_air_max       | 15.4               | float   | Mean month air temperature warmest month [C]
start_year      | 2000              | long     | begining of the simulation time [year]
end_year        | 2000                | long     | end of the simulation time [year]
dt              | 1.0                      | float    | Timestep for permafrost process [year]
#===============================================================================
# Output saved for user
#===============================================================================
fn_out_filename       | frostnumber_output.dat   | string       | Name of output file
```

**Comment by mdpiper** [comment]
Another way of thinking of this. We should be able to do this:

```
In [3]: x.initialize('frostnumber_model.cfg')
In [4]: x.update_until(x.get_end_time())
Warning: update_until year is less than current year
  no update run
```

**Comment by sc0tts** [comment]
Yes, this makes sense.

In the uncoupled case -- e.g. a single model -- I think it's fine to set
the values for the first time (timestep = 0) at initialization, and then
not allow time to progress past that.

So your second option above, don't throw the exception and don't change the
value.

In a couples scenario, things get trickier. We may even need to trick WMT
with something like update_until(timestep = current_timestep)

The problem arises even with FNGeo. If the calculation of FN depends on
temperature which comes from cruAKtemp, how do I get the temperature for
the first (0th) timestep?  If I initialize FNGeo first, then the framework
simply doesn't know the temperature yet. But if I initialize cruAKtemp
first, the it doesn't have anyplace (in the FNGeo model that hasn't loaded
yet) to set the temperature.

**Comment by wk1984** [comment]
If we only run the model once, the "update" is not needed?

**Comment by mdpiper** [comment]
> If we only run the model once, the "update" is not needed?

@wk1984 As far as I understand, this is the way the CSDMS Modeling Framework works. For example, to solve a really simple model

    y = x + 1

the initialize step is used to set the value of *x* at `t = 0`, then the update step is used to solve for *y* at `t = 1`.

**Comment by mdpiper** [comment]
> The problem arises even with FNGeo. If the calculation of FN depends on
temperature which comes from cruAKtemp, how do I get the temperature for
the first (0th) timestep?  If I initialize FNGeo first, then the framework
simply doesn't know the temperature yet. But if I initialize cruAKtemp
first, the it doesn't have anyplace (in the FNGeo model that hasn't loaded
yet) to set the temperature.

@sc0tts CRUAKTemp is the provides port for FNGeo, so it's initialized and updated first. PyMT handles it like this (some psuedocode):

```
from pymt.components import FrostNumberGeoModel, CRUAKTemp

f = FrostNumberGeoModel(), c = CRUAKTemp()
c.initialize('fileb.cfg')
f.initialize('filea.cfg')
while f.get_current_time() < f.get_end_time():
   temp = c.get_value('temperature__surface')
   f.set_value('temperature__surface', temp)
   c.update()
   f.update()
c.finalize(), f.finalize()
```

**Comment by sc0tts** [comment]
How accurate is this pseudocode?

I ask because looking at this, it seems like the last step in update()
should be to update the time, rather than the first.

**Comment by mdpiper** [comment]
@sc0tts I think I have a compromise. If I remove the "else" clause from the *update* method in **frost_number.py**, I can call the BMI *update* method at least once from the CMF. Alternately, as you hint at above, `self.year` could be incremented at the end of *update* rather than at the beginning. 

Would either of these be acceptable?

**Comment by mdpiper** [comment]
@sc0tts @wk1984 I spoke with @mcflugen this morning. I'm being too rigid in arguing that the CMF requires a model to update at least once. This is philosophically incorrect. However, practically, it's true: the CMF, as currently written, requires an update step. There's nothing stopping us from changing the CMF, except that I'd rather not so close to the meeting.

**Comment by sc0tts** [comment]
Agree: don't change the CMF now!  Too much risk of unintended consequences.

Updating the time at the end of the update() will work.  For now, at least.

Actually, this is the type of issue that Kevin was talking about: Unless
the framework keeps track of the time, in which case "update()" means "give
me your values for the current framework time", things will get...hard.
But this is an issue for June.

**Comment by mdpiper** [comment]
Thanks, @sc0tts. I'll leave this issue open for a while for reflection, but I'll move ahead with the workaround.

**Comment by sc0tts** [comment]
I can change the details of frost_number.py's update() after #42 is accepted, or you can do it as part of the acceptance process

**Comment by mdpiper** [comment]
@sc0tts I implemented the workaround in #43, which has been merged back into master.

**Comment by sc0tts** [comment]
Cool.  So strictly speaking, this issue is closed, though it there may be a more general issue about time and updating in coupled components.

**Comment by sc0tts** [comment]
When I started updating the time after calculating the values, my tests got all messed up.  So I can up with a different solution:  Allow update_frac() to take "0.0" as a value and calculate the value without updating the timestep.

Also, if the time exceeds the last time, it simply doesn't update.  (Though it prints something to say that.)

This way, the model's current state has values that represent the current time.

Will that be acceptable?

**Comment by mdpiper** [comment]
Note: the workaround discussed above is implemented in #48, not #43.

**Comment by sc0tts** [comment]
Current code calls update().  Either indirectly with update_frac(0) if coupled with another component, or without advancing the final timestep.
